pest/src/parser_state.rs

```diff
@@ -42,61 +48,151 @@ pub struct ParserState<'i, R: RuleType> {
-        atomicity: Atomicity::NonAtomic,
-        stack: vec![]
-    };
-
-    if f(&mut state, Position::from_start(input)).is_ok() {
-        let len = state.queue.len();
-        Ok(pairs::new(Rc::new(state.queue), input.as_bytes(), 0, len))
-    } else {
-        state.pos_attempts.sort();
-        state.pos_attempts.dedup();
-        state.neg_attempts.sort();
-        state.neg_attempts.dedup();
-
-        Err(Error::ParsingError {
-            positives: state.pos_attempts,
-            negatives: state.neg_attempts,
-            // All attempted positions were legal.
-            pos: unsafe { position::new(input.as_bytes(), state.attempt_pos) }
-        })
+    let state = ParserState::new(input);
+
+    match f(state) {
+        Ok(s) => {
+            let len = s.queue.len();
+            Ok(pairs::new(Rc::new(s.queue), input.as_bytes(), 0, len))
+        }
+        Err(mut s) => {
+            s.pos_attempts.sort();
+            s.pos_attempts.dedup();
+            s.neg_attempts.sort();
+            s.neg_attempts.dedup();
+
+            Err(Error::ParsingError {
+                positives: s.pos_attempts.clone(),
+                negatives: s.neg_attempts.clone(),
+                // All attempted positions were legal.
+                pos: unsafe { position::new(input.as_bytes(), s.attempt_pos) }
+            })
+        }
     }
 }
 
 impl<'i, R: RuleType> ParserState<'i, R> {
-    /// Wrapper needed to generate tokens.
+
+    /// Allocates a fresh `ParserState` object to the heap and returns the owned `Box`. This `Box`
+    /// will be passed from closure to closure based on the needs of the specified `Parser`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use pest;
+    ///
+    /// let input = "";
+    /// let state: Box<pest::ParserState<&str>> = pest::ParserState::new(input);
+    /// ```
+    pub fn new(input: &'i str) -> Box<Self> {
+        Box::new(ParserState {
+            position: Position::from_start(input),
+            queue: vec![],
+            lookahead: Lookahead::None,
+            pos_attempts: vec![],
+            neg_attempts: vec![],
+            attempt_pos: 0,
+            atomicity: Atomicity::NonAtomic,
+            stack: Stack::new()
+        })
+    }
+
+    /// Returns a reference to the current `Position` of the `ParserState`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use pest;
+    ///
+    /// # #[allow(non_camel_case_types)]
+    /// # #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+    /// enum Rule {
+    ///     ab
+    /// }
+    ///
+    /// let input = "ab";
+    /// let mut state: Box<pest::ParserState<Rule>> = pest::ParserState::new(input);
+    /// let position = state.get_position();
+    /// assert_eq!(position.pos(), 0);
+    /// ```
+    pub fn get_position(&self) -> &Position<'i> {
+        &self.position
+    }
+
+    /// Returns a clone of the current `Position` of the `ParserState`. Ideally, this will likely
```

Comment:
Not completely sure, but maybe a better API here would be to have a more functional approach. Instead of imperatively cloning and reassigning the position, we could use a `map_position<F>(f: F) where F: FnOnce(Position) -> Position`.

Reply:
Any thoughts on this one?